### ISSUE TITLE
`server_requests_spec.lua` hangs

### ISSUE THREAD

**Issue by mhinz** [neovim/neovim#2430]
676133a added new tests that hang for me.

The culprit is this line: https://github.com/neovim/neovim/blob/master/test/functional/api/server_requests_spec.lua#L131

I can reproduce it with:

```vim
let vim = rpcstart('nvim', ['--embed'])
call rpcnotify(vim, 'vim_set_current_line', 'SOME TEXT')
call rpcrequest(vim, 'vim_eval', '0')
```

**Comment by splinterofchaos** [comment]
@mhinz, could you apply this patch and tell me if nvim outputs anything?
```patch
diff --git a/src/nvim/msgpack_rpc/channel.c b/src/nvim/msgpack_rpc/channel.c
index b671b8b..2cfec0f 100644
--- a/src/nvim/msgpack_rpc/channel.c
+++ b/src/nvim/msgpack_rpc/channel.c
@@ -398,10 +398,15 @@ static void parse_msgpack(RStream *rstream, void *data, bool eof)
       }
       msgpack_unpacked_destroy(&unpacked);
       // Bail out from this event loop iteration
-      goto end;
+    } else {
+      handle_request(channel, &unpacked.data);
     }
+  }
 
-    handle_request(channel, &unpacked.data);
+  if (result == MSGPACK_UNPACKER_EXTRA_BYTES) {
+    EMSG("EXTRA");
+  } else if (result == MSGPACK_UNPACKER_CONTINUE) {
+    EMSG("CONT");
   }
 
   if (result == MSGPACK_UNPACK_NOMEM_ERROR) {
```

**Comment by splinterofchaos** [comment]
@mhinz and I discussed this a bit on gitter. Posting here for documentation purposes.

From the [`.nvimlog`](https://gist.github.com/mhinz/e8617425c72efd018595) produced by running the above script (edited and annotated with my best understanding): 
```
2015/04/16 00:00:36 [debug @ log_server_msg:822] parent - [msgpack-rpc] nvim -> client(1) [request]      [0, 1, "vim_eval", ["0"]]
2015/04/16 00:00:36 [debug @ loop:180] parent - Enter event loop
2015/04/16 00:00:36 [debug @ loop:182] parent - Exit event loop
2015/04/16 00:00:36 [debug @ loop:180] parent - Enter event loop
-- Parent enters event loop, waiting for the child to respond.

2015/04/16 00:00:36 [debug @ rbuffer_produced:76] child - Received 17 bytes from RStream(0x78fc060)
2015/04/16 00:00:36 [debug @ parse_msgpack:369] child - Feeding the msgpack parser with 17 bytes of data from RStream(0x78fc060)
2015/04/16 00:00:36 [debug @ log_client_msg:833] child - [msgpack-rpc] client(1) -> nvim [request]      [0, 1, "vim_eval", ["0"]]
-- Child receives the request and defers it.
2015/04/16 00:00:36 [debug @ loop:182] child - Exit event loop
2015/04/16 00:00:36 [debug @ loop:180] child - Enter event loop
-- Child enters the event loop and freezes.
-- After 14 seconds, @mhinz tries to kill the process.
2015/04/16 00:00:40 [debug @ rbuffer_produced:76] parent - Received 1 bytes from RStream(0x5501060)
-- Parent regains conciousness.
2015/04/16 00:00:40 [debug @ rbuffer_produced:76] parent - Received 1 bytes from RStream(0x0)
2015/04/16 00:00:40 [debug @ loop:182] parent - Exit event loop
2015/04/16 00:00:40 [debug @ loop:180] parent - Enter event loop
2015/04/16 00:00:50 [debug @ loop:182] child - Exit event loop
2015/04/16 00:00:50 [debug @ rbuffer_produced:76] parent - Received 37 bytes from RStream(0x5501380)
2015/04/16 00:00:50 [error @ job_err:346] parent - Channel 1 stderr: Vim: Caught deadly signal 'SIGTERM'
 -- Parent receives a message that the client got sent a SIGTERM
2015/04/16 00:00:50 [debug @ loop:182] parent - Exit event loop
2015/04/16 00:00:50 [debug @ loop:180] parent - Enter event loop
2015/04/16 00:00:50 [debug @ rbuffer_produced:76] parent - Received 15 bytes from RStream(0x5501380)
2015/04/16 00:00:50 [error @ job_err:346] parent - Channel 1 stderr: Vim: Finished.
2015/04/16 00:00:50 [debug @ loop:182] parent - Exit event loop
2015/04/16 00:00:50 [debug @ loop:180] parent - Enter event loop
2015/04/16 00:00:50 [debug @ loop:180] child - Enter event loop
2015/04/16 00:00:50 [debug @ loop:182] child - Exit event loop
2015/04/16 00:00:50 [debug @ loop:180] child - Enter event loop
2015/04/16 00:00:52 [debug @ loop:182] child - Exit event loop
2015/04/16 00:00:52 [debug @ loop:180] child - Enter event loop
2015/04/16 00:00:52 [debug @ loop:182] child - Exit event loop
2015/04/16 00:00:52 [debug @ loop:180] child - Enter event loop
2015/04/16 00:00:52 [debug @ loop:182] child - Exit event loop
2015/04/16 00:00:52 [debug @ log_server_msg:822] child - [msgpack-rpc] nvim -> client(1) [response]     [1, 1, nil, 0]
-- Ten seconds after the SIGTERM, the child composes its response.
2015/04/16 00:00:52 [debug @ rbuffer_produced:76] parent - Received 5 bytes from RStream(0x5501330)
2015/04/16 00:00:52 [debug @ parse_msgpack:369] parent - Feeding the msgpack parser with 5 bytes of data from RStream(0x5501330)
2015/04/16 00:00:52 [debug @ log_client_msg:833] parent - [msgpack-rpc] client(1) -> nvim [response]     [1, 1, nil, 0]
-- Parent receives the message, but it's too late.
```
I think the child is stuck in an infinite loop, maybe due to `event_poll(0)`, and the parent is stuck, waiting on the child.

**Comment by mhinz** [comment]
Is there any way to _leave out_ certain tests? This problem basically prevents me from using `make test`.

**Comment by splinterofchaos** [comment]
Sorry, not that I'm aware of. (Maybe `TEST_FILTER` accepts negations?)  I've spotted some other problems this might be the source of.

@tarruda, any thoughts on this?

**Comment by Numkil** [comment]
Just wanted to let you guys know that tests aren't hanging for me any more since the latest commits.

**Comment by justinmk** [comment]
Guess it was the `-u NONE` issue that @mhinz noticed. Thanks for followup @Numkil

**Comment by mhinz** [comment]
Actually it was fixed in: 42de617

If you didn't have the `neovim` module installed, nvim would prompt for it at start.. and that confused the tests, since they needed an addtional `<cr>`.

But yes, fixed. :]
